File: c++/cote/GJM/합이_0에_가장_가까운_구간.cpp
#include <iostream>
#include <random>
#include <vector>
#include <map>

#define TEST_CASES 10
#define N 1000
#define MAX_VAL 987654321

using namespace std;


vector<pair<int,int>> partSum(vector<int>& l){

	vector<pair<int,int>> ret(l.size());

	ret[0] = make_pair(l[0], 0);
	for(int i=1; i<l.size(); i++)
		ret[i] = make_pair(ret[i-1].first + l[i], i);

	return ret;
}


pair<int,int> solve(vector<int>& l){
	int min_val = MAX_VAL;
	int cur = 0;
	vector<pair<int,int>> ret = partSum(l);
	sort(begin(ret), end(ret), [](auto a, auto b){ return a.first < b.first; });


	for(int i= 1; i<ret.size(); i++){
		if(abs(ret[i].first - ret[i-1].first) < min_val){
			min_val = abs(ret[i].first - ret[i-1].first);
			cur = i;
		}
	}
	
	int first = min(ret[cur].second, ret[cur-1].second) + 1;
	int second = max(ret[cur].second, ret[cur-1].second);
	
	return make_pair(first, second);
}

int main(void){

	default_random_engine gen(43);

	for(int it=0; it<TEST_CASES; it++){
		normal_distribution<float> dis(0, 400);

		vector<int> l;

		for(int i=0; i<N; i++)
			l.push_back(floor(dis(gen)));

		pair<int,int> result = solve(l) ;
		cout << result.first << "," << result.second<< endl;

	}


	return 0;
}

File: go/cote/GJM/합이_0에_가장_가까운_구간.go
package main

import (
	"fmt"
	"math"
	"math/rand"
	"sort"
)

const (
	TEST_CASES = 10
	N          = 1000
	MAX_VAL    = 987654321
)

type Entry struct {
	Sum   int
	Index int
}

func minmax(i, j int) (int, int) {
	if i < j {
		return i, j
	}
	return j, i
}

func partialSum(l []int) []Entry {
	ret := make([]Entry, len(l))

	ret[0] = Entry{Sum: l[0], Index: 0}

	for i := 1; i < len(l); i++ {
		ret[i] = Entry{Sum: ret[i-1].Sum + l[i], Index: i}
	}

	return ret
}

func solve(l []int) (int, int) {

	ps := partialSum(l)

	sort.Slice(ps, func(i, j int) bool {
		return ps[i].Sum < ps[j].Sum
	})

	min, cur := MAX_VAL, 0

	for i := 1; i < len(ps); i++ {
		if math.Abs(float64(ps[i].Sum-ps[i-1].Sum)) < float64(min) {
			min = int(math.Abs(float64(ps[i].Sum - ps[i-1].Sum)))
			cur = i
		}
	}
	start, end := minmax(ps[cur].Index, ps[cur-1].Index)

	return start + 1, end
}

func main() {
	random := rand.New(rand.NewSource(43))

	for it := 0; it < TEST_CASES; it++ {
		n := random.Intn(N-2) + 2
		l := make([]int, n)

		for i := 0; i < n; i++ {
			l[i] = random.Intn(800) - 400
		}

		start, end := solve(l)
		fmt.Println(start, end)
	}
}
